### PR TITLE
Ensure the C++ UIA rate limited event handler flusher thread is definitely terminated when NVDA shutts down

### LINKED ISSUE #16072: Portable version of NVDA starts very slowly and is unusable
When I try to start a portable copy of NVDA (I have used many Alpha and Beta versions since December 2023) it may start the first time, but when I exit the program and start it again, NVDA tries to start but this happens very slowly, and finally I have to boot the computer, if I don't manage to kill the NVDA process. Sometimes I can't kill the NVDA process and reboot seems to be the only solution. NVDA doesn't react on key presses and my Braille display (Braillex EL 80c) is empty.

I have never used installed versions of NVDA, and everything has gone correctly until last December when I moved from 32 bit  Windows 7 to 64 bit Windows 11 Pro. I am using the same setting files which I used in Windows 7. I put the portable copy to C:nvda (beta version) and C:\nvda2 (Alpha version) and copied the one userconfig folder to these NVDA folders. Is it possible that my problems have something to do with userconfig folder's content, normal text-based files?
### Steps to reproduce:
Put the portable NVDA to a folder, i.e. C:\nvda, and start and quit it a few times.
### Actual behavior:

NVDA starts very slowly, the speech synthesizer says "Starting NVDA. Please wait..." and nothing happens after this.
### Expected behavior:

NVDA starts normally and it can be used normally.
### NVDA logs, crash dumps and other attachments:
Here are some lines from the nvda.log file? Should I post the whole file? I'm not sure

Initializing MathPlayer
WARNING - mathPres.initialize (08:06:34.408) - MainThread (12452):
MathPlayer 4 not available
DEBUGWARNING - core.main (08:06:34.408) - MainThread (12452):
Slow starting core (10.44 sec)
IO - speech.speech.speak (08:06:34.408) - MainThread (12452):
Speaking ['Loading NVDA. Please wait...']
DEBUGWARNING - characterProcessing._getSpeechSymbolsForLocale (08:06:34.408) - MainThread (12452):
No CLDR data for locale fi_FI
DEBUGWARNING - characterProcessing._getSpeechSymbolsForLocale (08:06:34.408) - MainThread (12452):
No symbol data for locale fi_FI
INFO - core.main (08:06:34.497) - MainThread (12452):
Using wx version 4.2.1 msw (phoenix) wxWidgets 3.2.2.1 with six version 1.16.0
DEBUG - core.main (08:06:34.497) - MainThread (12452):
Initializing braille input
INFO - brailleInput.initialize (08:06:34.497) - MainThread (12452):
Braille input initialized
DEBUG - core.main (08:06:34.497) - MainThread (12452):
Initializing braille
INFO - braille.initialize (08:06:34.497) - MainThread (12452):
Using liblouis version 3.28.0
INFO - braille.initialize (08:06:34.497) - MainThread (12452):
Using pySerial version 3.5
DEBUG - autoSettingsUtils.autoSettings.AutoSettings._registerConfigSaveAction (08:06:34.497) - MainThread (12452):
registering pre_configSave action: <class 'brailleDisplayDrivers.papenmeier.BrailleDisplayDriver'>
INFO - brailleDisplayDrivers.papenmeier.BrailleDisplayDriver.connectUSB (08:06:34.559) - MainThread (12452):
connectUSB success
INFO - brailleDisplayDrivers.papenmeier.BrailleDisplayDriver.__init__ (08:06:34.617) - MainThread (12452):
Found EL80c
DEBUG - autoSettingsUtils.autoSettings.AutoSettings._loadSpecificSettings (08:06:34.617) - MainThread (12452):
loading braille papenmeier
INFO - braille.BrailleHandler._setDisplay (08:06:34.617) - MainThread (12452):
Loaded braille display driver 'papenmeier', current display has 80 cells.
DEBUG - core.main (08:06:34.617) - MainThread (12452):
Initializing vision
DEBUG - core.main (08:06:34.617) - MainThread (12452):
Initializing displayModel
DEBUG - core.main (08:06:34.617) - MainThread (12452):
Initializing GUI
DEBUG - config.isInstalledCopy (08:06:34.627) - MainThread (12452):
Unable to find isInstalledCopy registry key RegistryKey.INSTALLED_COPY- this is not an installed copy.
#### NVDA installed/portable/running from source:
Portable
#### NVDA version:
2024.1beta4
#### Windows version:
Microsoft Windows 11 Professional Version 23H2 (OS Build 22631.3007), System Type x64
#### Name and version of other software in use when reproducing the issue:
No other software
#### Other information about your system:
Lenovo ThinkStation P3 Tower with 16 GB RAM
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Alphas and Betas since December 2023 with same results
#### If NVDA add-ons are disabled, is your problem still occurring?
Yes
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
Fixes #16072

### Summary of the issue:
If any background threads remain after NVDA exits, the NVDA process will stay alive until those background threads complete. For Python threads, setting daemon to true on the thread causes Python to kill off the thread when the main thread exits. However, Python is unaware of C++ threads.
One particular C++ thread which has proven to keep NvDA alive for at least one person (see #16072) is the UIA rate limited event handler flusher thread.
We should try much harder to ensure that this thread is terminated during NVDA exit so that it does not keep the NVDA process alive after the main thread exits.

### Description of user facing changes
The NVDA process is less likely to stay around after NvDA has exited, which means that NVDA should no longer freeze on restart and NvDA updates should be less likely to fail due to NVDA still running.

### Description of development approach
* UIA rate limited event handler: improve logic and logging of the flusher thread function to make it clearer as to why the thread will wake for flushes or when it should stop.
* Add the ability to terminate UIA rate limited event handler's flusher thread upon request via a new terminate API call. This allows NvDA to ensure that the flusher thread can be terminated, even if it may not release the rate limited event handler COM object reference for some reason.
* UIAHandler.terminate: specifically terminate the rate limited event handler from NvDA's main thread, before trying to terminate the UIA MTA thread. This ensures that the c++ UIA rate limitied event handler flusher thread is definitely terminated, even if there may be further errors terminating the UIA MTA Python thread.

### Testing strategy:
* Compiled NVDAHelperLocal with debug logging enabled.
* Ensured that NVDA's Enhanced UIA event processing advanced option is turned on.
* Started and quit NvDA several times.
* Confirmed in NvDA's log that the UIA rate limited event hander flusher thread correctly woke and terminated durng UIAHandler.terminate.
* [ ] The reporter of #16072 would still need to test and confirm that this change stops the issues they are experiencing.

### Known issues with pull request:
None known.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

File: source/UIAHandler/__init__.py [hunk 1]
```diff
@@ -440,6 +441,15 @@ def __init__(self):
 			raise self.MTAThreadInitException
 
 	def terminate(self):
+		# Terminate the rate limited event handler if it exists.
+		# We must do this from the main thread to totally ensure that the thread is terminated,
+		# As this is a c++ thread so Python canot kill it off at process exit.
```
Comment (lukaszgo1) [inline source/UIAHandler/__init__.py:446]:
```suggestion
		# As this is a c++ thread so Python cannot kill it off at process exit.
```

**Comment by lukaszgo1** [comment]
I assume it is not clear why the thread fails to end normally for the reporter of #16072 ? Would it be possible to debug this further, perhaps with additional logging? Obviously for 2024.1 it is important the thread gets terminated and does not block NVDA's main process, but perhaps for next release understanding the real cause may prove useful.

**Comment by michaelDCurran** [comment]
I do not see any error in the provided logs so far unfortunately.

Of course if I do get more information which suggests further 
investigation I shall do so.

Remember that no one is able to reproduce this except for the reporter 
of #16072. So for now it is just trying to take precautions. Though we 
can at least be sure that we never want a c++ thread to remain after the 
main thread exits for anyone.

File: nvdaHelper/local/UIAEventLimiter/api.cpp [hunk 2]
```diff
@@ -30,5 +33,20 @@ HRESULT rateLimitedUIAEventHandler_create(IUnknown* pExistingHandler, RateLimite
...
+	activeRateLimitedEventHandlers.insert(*ppRateLimitedEventHandler);
+	return S_OK;
+}
+
+// @brief Terminates a RateLimitedEventHandler instance's flusher thread and removes it from the activeRateLimitedEventHandlers set.
+// @param pRateLimitedEventHandler the RateLimitedEventHandler instance to terminate.
+// @note This function will block until the RateLimitedEventHandler's flusher thread has terminated. 
+// @return S_OK on success or a failure code otherwise
+HRESULT rateLimitedUIAEventHandler_terminate(RateLimitedEventHandler* pRateLimitedEventHandler) {
+	if(activeRateLimitedEventHandlers.find(pRateLimitedEventHandler) == activeRateLimitedEventHandlers.end()) {
```
Comment (seanbudd) [inline nvdaHelper/local/UIAEventLimiter/api.cpp:45]:
```suggestion
	if (activeRateLimitedEventHandlers.find(pRateLimitedEventHandler) == activeRateLimitedEventHandlers.end()) {
```

File: nvdaHelper/local/UIAEventLimiter/rateLimitedEventHandler.cpp [hunk 3]
```diff
@@ -66,25 +66,38 @@ void RateLimitedEventHandler::flusherThreadFunc(std::stop_token stopToken) {
...
 	do { // thread main loop
-		LOG_DEBUG(L"flusherThreadFunc sleeping...");
+		bool needsFlush = false;
 		{ std::unique_lock lock(m_mtx);
 			// sleep until a flush is needed or this thread should stop.
+			LOG_DEBUG(L"flusherThreadFunc sleeping...");
 			m_flushConditionVar.wait(lock, [this, stopToken](){
-				return this->m_needsFlush || stopToken.stop_requested();
+				LOG_DEBUG(L"FLUSHER THREAD notified, CHECKING WAKE CONDITION");
+				if(this->m_needsFlush ) {
```
Comment (seanbudd) [inline nvdaHelper/local/UIAEventLimiter/rateLimitedEventHandler.cpp:79]:
```suggestion
				if (this->m_needsFlush) {
```

File: nvdaHelper/local/UIAEventLimiter/rateLimitedEventHandler.cpp [hunk 4]
```diff
@@ -66,25 +66,38 @@ void RateLimitedEventHandler::flusherThreadFunc(std::stop_token stopToken) {
...
+					LOG_DEBUG(L"Will wake as needsFlush is true");
+					return true;
+				} else if(stopToken.stop_requested()) {
```
Comment (seanbudd) [inline nvdaHelper/local/UIAEventLimiter/rateLimitedEventHandler.cpp:82]:
```suggestion
				} else if (stopToken.stop_requested()) {
```

File: nvdaHelper/local/UIAEventLimiter/rateLimitedEventHandler.cpp [hunk 5]
```diff
@@ -66,25 +66,38 @@ void RateLimitedEventHandler::flusherThreadFunc(std::stop_token stopToken) {
...
+					return true;
+				}
+				LOG_DEBUG(L"Spurious wake, will sleep again");
+				return false;
 			});
 			LOG_DEBUG(L"flusherThread woke up");
-			if(stopToken.stop_requested()) {
-				LOG_DEBUG(L"flusherThread returning as stop requested"); 
-				return;
+			if(this->m_needsFlush ) {
```
Comment (seanbudd) [inline nvdaHelper/local/UIAEventLimiter/rateLimitedEventHandler.cpp:90]:
```suggestion
			if (this->m_needsFlush) {
```

File: nvdaHelper/local/UIAEventLimiter/rateLimitedEventHandler.cpp [hunk 6]
```diff
@@ -66,25 +66,38 @@ void RateLimitedEventHandler::flusherThreadFunc(std::stop_token stopToken) {
...
+				m_needsFlush = false;
+				needsFlush = true;
 			}
-			m_needsFlush = false;
 		} // m_mtx released here.
-		flushEvents();
+		if(needsFlush) {
```
Comment (seanbudd) [inline nvdaHelper/local/UIAEventLimiter/rateLimitedEventHandler.cpp:95]:
```suggestion
		if (needsFlush) {
```

**Review by seanbudd — DISMISSED** [review]
just minor lint fixes

**Review by seanbudd — APPROVED** [review]
just minor lint fixes

**Pull request merged**